platform/iphone/view_controller.mm

```diff
@@ -92,37 +90,119 @@ - (BOOL)shouldAutorotate {
+}
+
+- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
+	UITouch *touch = [touches anyObject];
+
+	CGPoint touchPoint = [self scaledPoint:[touch locationInView:self.view]];
+	OSIPhone::get_singleton()->touch_press([self.activeTouches indexOfObject:touch], touchPoint.x, touchPoint.y, false, false);
+
+	[self.activeTouches removeObject:touch];
+}
+
+- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event {
+	UITouch *touch = [touches anyObject];
+
+	OSIPhone::get_singleton()->touches_cancelled();
+
+	[self.activeTouches removeObject:touch];
+}
+
+- (CGPoint)scaledPoint:(CGPoint)point {
+	return CGPointMake(point.x * self.view.contentScaleFactor, point.y * self.view.contentScaleFactor);
+}
+
+#pragma mark - Keyboard Input
+- (BOOL)canBecomeFirstResponder {
+	return YES;
+}
+
+- (void)keyboardOnScreen:(NSNotification *)notification {
+	NSValue *value = notification.userInfo[UIKeyboardFrameEndUserInfoKey];
+	CGRect frame = [value CGRectValue];
+	const CGFloat kScaledHeight = frame.size.height * [UIScreen mainScreen].nativeScale;
+	OSIPhone::get_singleton()->set_virtual_keyboard_height(kScaledHeight);
+}
+
+- (void)keyboardHidden:(NSNotification *)notification {
+	OSIPhone::get_singleton()->set_virtual_keyboard_height(0);
+}
+
+- (void)deleteBackward {
+	if (keyboard_text.length())
+		keyboard_text.erase(keyboard_text.length() - 1, 1);
+	OSIPhone::get_singleton()->key(KEY_BACKSPACE, true);
+}
+
+- (BOOL)hasText {
+	return keyboard_text.length() ? YES : NO;
+}
+
+- (void)insertText:(NSString *)p_text {
+	String character;
+	character.parse_utf8([p_text UTF8String]);
+	keyboard_text = keyboard_text + character;
+	OSIPhone::get_singleton()->key(character[0] == 10 ? KEY_ENTER : character[0], true);
+	printf("inserting text with character %i\n", character[0]);
+}
+
+#pragma mark - Game Center
 - (void)gameCenterViewControllerDidFinish:(GKGameCenterViewController *)gameCenterViewController {
+#ifdef GAME_CENTER_ENABLED
 	//[gameCenterViewController dismissViewControllerAnimated:YES completion:^{GameCenter::get_singleton()->game_center_closed();}];//version for signaling when overlay is completely gone
 	GameCenter::get_singleton()->game_center_closed();
 	[gameCenterViewController dismissViewControllerAnimated:YES completion:nil];
-}
 #endif
+}
 
 @end
+
+#pragma mark - Bridged Text Input
+void _show_keyboard(String p_existing) {
+	keyboard_text = p_existing;
+	NSLog(@"Show keyboard");
+	[_instance becomeFirstResponder];
+}
+
+void _hide_keyboard() {
+	NSLog(@"Hide keyboard and clear text");
+	[_instance resignFirstResponder];
+	keyboard_text = "";
+}
+
+Rect2 _get_ios_window_safe_area(float p_window_width, float p_window_height) {
+	UIEdgeInsets insets = UIEdgeInsetsMake(0, 0, 0, 0);
+	if ([_instance.view respondsToSelector:@selector(safeAreaInsets)]) {
+		insets = [_instance.view safeAreaInsets];
```

Comment:
This needs an availability check apparently:
```
platform/iphone/gl_view.mm:84:23:{84:23-84:38}: warning: 'safeAreaInsets' is only available on iOS 11.0 or newer [-Wunguarded-availability-new,2]
                insets = [_instance safeAreaInsets];
                                     ^~~~~~~~~~~~~~~
/Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS11.4.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIView.h:263:45: note: 'safeAreaInsets' has been explicitly marked partial here [2]
 @property (nonatomic,readonly) UIEdgeInsets safeAreaInsets API_AVAILABLE(ios(11.0),tvos(11.0));
                                             ^
platform/iphone/gl_view.mm:84:23:{84:23-84:38}: note: enclose 'safeAreaInsets' in an @available check to silence this warning [2]
                insets = [_instance safeAreaInsets];
                                     ^~~~~~~~~~~~~~~
fix-it:"platform/iphone/gl_view.mm":{84:3-84:3}:"if (@available(iOS 11.0, *)) {\n\t\t    "
fix-it:"platform/iphone/gl_view.mm":{84:39-84:39}:"\n\t\t} else {\n\t\t    // Fallback on earlier versions\n\t\t}"
```
(Warning for before this PR, but from what I can see it should still be valid.)